src/Symfony/Bridge/Doctrine/Form/ChoiceList/DoctrineChoiceLoader.php

```diff
@@ -59,81 +52,63 @@ public function __construct(ObjectManager $manager, string $class, IdReader $idR
     /**
      * {@inheritdoc}
      */
-    public function loadChoiceList(callable $value = null)
+    protected function loadChoices(): iterable
     {
-        if ($this->choiceList) {
-            return $this->choiceList;
-        }
-
-        $objects = $this->objectLoader
+        return $this->objectLoader
             ? $this->objectLoader->getEntities()
-            : $this->manager->getRepository($this->class)->findAll();
-
-        return $this->choiceList = new ArrayChoiceList($objects, $value);
+            : $this->manager->getRepository($this->class)->findAll()
+        ;
     }
 
     /**
-     * {@inheritdoc}
+     * This method is here for legacy.
+     *
+     * To remove in 6.0
      */
-    public function loadValuesForChoices(array $choices, callable $value = null)
+    protected function doLoadValuesForChoices(array $choices): array
     {
-        // Performance optimization
-        if (empty($choices)) {
-            return [];
-        }
-
         // Optimize performance for single-field identifiers. We already
         // know that the IDs are used as values
-        $optimize = $this->idReader && (null === $value || \is_array($value) && $value[0] === $this->idReader);
-
         // Attention: This optimization does not check choices for existence
-        if ($optimize && !$this->choiceList) {
-            $values = [];
-
+        if ($this->idReader) {
+            @trigger_error(sprintf('Not defining explicitly the IdReader as value callback when query can be optimized has been deprecated in 5.1. Don\'t pass the IdReader to "%s" or define the "choice_value" option instead.', __CLASS__), E_USER_DEPRECATED);
             // Maintain order and indices of the given objects
+            $values = [];
             foreach ($choices as $i => $object) {
                 if ($object instanceof $this->class) {
-                    // Make sure to convert to the right format
-                    $values[$i] = (string) $this->idReader->getIdValue($object);
+                    $values[$i] = $this->idReader->getIdValue($object);
```

Comment:
there is no guarantee that userland implementations always return a string, so we should keep the cast, isn't it? better reduce the diff also